sysl2/sysl/integration/deps.go

```diff
@@ -0,0 +1,304 @@
+		}
+	}
+
+	return integrations
+}
+
+func (ds *DependencySet) ResolveDependencies(module *sysl.Module) {
+	apps := module.GetApps()
+	for appname, app := range apps {
+		for epname, endpoint := range app.GetEndpoints() {
+			ds.resolveStatementDependencies(endpoint.GetStmt(), apps, appname, epname)
+		}
+	}
+
+	if len(ds.errors) > 0 {
+		panic(fmt.Sprintf("broken deps:\n  %s", strings.Join(ds.errors, "\n")))
+	}
+}
+
+func (ds *DependencySet) resolveStatementDependencies(stmts []*sysl.Statement, apps map[string]*sysl.Application, appname, epname string) {
+	for _, stat := range stmts {
+		switch c := stat.GetStmt().(type) {
+		case *sysl.Statement_Call:
+			var errStr string
+			targetName := getAppName(c.Call.GetTarget())
+			targetApp := apps[targetName]
+			if targetApp == nil {
+				errStr = fmt.Sprintf("%s <- %s: calls non-existent app %s", appname, epname, targetName)
+				ds.errors = append(ds.errors, errStr)
+			} else {
+				isValid := !hasPattern("abstract", utils.MakeStrSetFromSpecificAttr("patterns", targetApp.GetAttrs()).ToSlice())
+				if !isValid {
+					panic(fmt.Sprintf("call target '%s' must not be ~abstract", targetName))
+				}
+				callEndpoint := c.Call.GetEndpoint()
+				if targetApp.GetEndpoints()[callEndpoint] == nil {
+					errStr = fmt.Sprintf(
+						"%s <- %s: calls non-existent endpoint %s -> %s", appname, epname, targetName, callEndpoint)
+					ds.errors = append(ds.errors, errStr)
+				} else {
+					selfApp := MakeAppElement(appname, epname)
+					targetApp := MakeAppElement(targetName, callEndpoint)
+					dep := MakeAppDependency(selfApp, targetApp)
+					ds.Deps[dep] = struct{}{}
+				}
+			}
+		case *sysl.Statement_Action, *sysl.Statement_Ret:
+			continue
+		case *sysl.Statement_Cond:
+			ds.resolveStatementDependencies(c.Cond.GetStmt(), apps, appname, epname)
+		case *sysl.Statement_Loop:
+			ds.resolveStatementDependencies(c.Loop.GetStmt(), apps, appname, epname)
+		case *sysl.Statement_LoopN:
+			ds.resolveStatementDependencies(c.LoopN.GetStmt(), apps, appname, epname)
+		case *sysl.Statement_Foreach:
+			ds.resolveStatementDependencies(c.Foreach.GetStmt(), apps, appname, epname)
+		case *sysl.Statement_Group:
+			ds.resolveStatementDependencies(c.Group.GetStmt(), apps, appname, epname)
+		case *sysl.Statement_Alt:
+			for _, choice := range c.Alt.GetChoice() {
+				ds.resolveStatementDependencies(choice.GetStmt(), apps, appname, epname)
+			}
+		default:
+			panic("No statement!")
+		}
+	}
+}
+
+func isSub(child, parent utils.StrSet) bool {
+	if len(child) == 0 && len(parent) == 0 {
+		return true
+	}
+	if len(parent) == 0 {
+		return false
+	}
+	if len(child) == 0 {
+		return true
+	}
+	return len(child.Difference(parent)) == 0
+}
+
+func hasPattern(s string, arr []string) bool {
+	m := buildStringBoolFilter(arr)
+
+	return m[s]
+}
+
+func toPattern(comp []string) string {
+	return fmt.Sprintf(`^(?:%s)(?: *::|$)`, strings.Join(comp, "|"))
+}
+
+func getAppName(appname *sysl.AppName) string {
```

Comment:
duplicate code. this function exists somewhere else too.

Reply:
move sysl/sysl2/sysl/helpers.go to sysl/sysl2/sysl/utils/helpers.go. and change packge import for parse.go, ValueUtil.go, listener_impl.go.

Reply:
@loki5265 i wouldnt want to do that. Just keep all the go files in the root. ints and seqs are not separate packages.
all of this is sysl package. so i would just keep everything as main.